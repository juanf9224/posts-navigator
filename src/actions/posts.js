
import { searchPosts } from '../services/post.service';

// SET_POSTS
export const setPosts = (posts) => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = () => {
  return (dispatch, getState) => {
    return searchPosts().then((res) => {
      dispatch(setPosts(res.data));
    });
  };
};

// EDIT_POST
export const editPostAction = (id, update) => ({
  type: 'EDIT_POST',
  id,
  update
});


