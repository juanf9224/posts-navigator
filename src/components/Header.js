import React from "react";


const Header = () => (
  <header className="header">
    <div className="header__title">
      <h1>Posts Navigator</h1>
    </div>
  </header>
);

export { Header as default };
