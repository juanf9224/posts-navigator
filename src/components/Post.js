import React, { useContext } from 'react';

import AutocompleteContext from '../context/autocomplete';

const Post = ({ post }) => {
  const { handleSuggestionClick, activeSuggestion } = useContext(AutocompleteContext);
  const { idx, id } = activeSuggestion;
  return (
    <li
      onClick={() =>
        handleSuggestionClick({ id: post.id, title: post.title })
      }
      className={
        idx > -1 && post.id === id ? "is-active post-item" : "post-item"
      }
    >
      <div className="post-item__title">{post.title}</div>
    </li>
  );
};
export { Post as default };