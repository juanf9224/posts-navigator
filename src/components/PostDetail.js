import React, { useContext } from "react";

const PostDetail = ({ title, body, setIsEdit }) => {
  return (
    <div className="card-layout">
      <div className="card-layout__content">
        <div className="card-layout__content__title">
          <h1>{title}</h1>
        </div>
        <div className="card-layout__content__body">
          <p>{body}</p>
        </div>
        <div className="card-layout__content__footer">
          <button class="btn-secondary" onClick={() => setIsEdit(true)}>
            Edit
          </button>
        </div>
      </div>
    </div>
  );
};

export { PostDetail as default };
