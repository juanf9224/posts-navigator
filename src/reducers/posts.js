export const postsReducerDefaultState = {
  items: [],
  error: ''
};

export default (state = postsReducerDefaultState, action) => {
  switch (action.type) {
    case 'EDIT_POST':
      return {
        ...state,
        items: state.items.map((post) => {
          if (post.id === action.id) {
            return {
              ...post,
              ...action.update,
            };
          }
          return post;
        })
      };
    case 'GET_POSTS':
      return {
        ...state,
      };
    case 'GET_POSTS_SUCCESS':
      return {
        ...state,
        items: action.payload,
        error: '',
      };
    case 'GET_POSTS_ERROR':
      return { ...state, error: action.payload };
    default:
      return state;
  }
};