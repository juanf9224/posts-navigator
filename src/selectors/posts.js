export default (posts, { text, pagination }) => {
  const {
    pageNumber,
    itemsPerPage
  } = pagination;
  const start = (pageNumber - 1) * itemsPerPage;
  const end = start + itemsPerPage;
  
  if (!text) return { posts: posts.slice(start, end), matching: null };

  const filteredPosts = posts && posts
    .filter((p) => p.title.includes(text));
  return { posts: filteredPosts.length < start ? filteredPosts : filteredPosts.slice(start, end), matching: filteredPosts };
};
