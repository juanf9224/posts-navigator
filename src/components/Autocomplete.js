/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect, Fragment } from 'react';
import { connect } from 'react-redux';
import Input from '@material-ui/core/Input';
import { makeStyles, createStyles } from "@material-ui/core";
import IconButton from '@material-ui/core/IconButton';
import ClearIcon from '@material-ui/icons/Clear';
import SearchIcon from '@material-ui/icons/Search';

import useDebounce from '../hooks/use-debounce';
import selectPosts from '../selectors/posts';
import { setTextFilter } from '../actions/filters';

const useStyles = makeStyles(() => createStyles({
  hidden: {
    display: 'none'
  },  
  searchInput: {
    width: '20rem',
    fontSize: '1.4rem'
  },
  btnIcon: {
    fontSize: '2rem'
  }
}));

export const Autocomplete = ({ filterResults, setTextFilter }) => {
  const [filterText, setFilterText] = useState('');
  // debounce hook to avoid too many requests
  const debouncedSearchTerm = useDebounce(filterText, 500);
  const classes = useStyles();
  const { posts } = filterResults;

  useEffect(() => {
    // Perform the post search with a custom hook to debounce
    // and avoid unnecessary queries
    setTextFilter(debouncedSearchTerm);
  }, [debouncedSearchTerm]);

  /**
   * Clears the input as well as all state related to the search
   */
  const clearSearch = () => {
    setFilterText('');
    setTextFilter('');
  };

  return (
    <Fragment>
      <Input
        id="search-input"
        placeholder="Search Post"
        value={filterText}
        onChange={(e) => {
          setFilterText(e.target.value);
        }}
        inputProps={{
          list: "item-list",
          "data-testid": "search-input",
          className: classes.searchInput,
        }}
      />
      <IconButton
        color="primary"
        aria-label="clear search"
        onClick={clearSearch}
        className={filterText.length < 1 ? classes.hidden : ""}
      >
        <ClearIcon />
      </IconButton>
      <IconButton
        color="primary"
        aria-label="clear search"
        className={filterText.length ? classes.hidden : ""}
      >
        <SearchIcon className={classes.btnIcon} />
      </IconButton>
      <datalist
        id="item-list"
        data-testid="autocomplete-data-list">
        {posts && posts.length ? (
          posts.map((p) => (
            <option key={p.id} value={p.title}>
              {p.title}
            </option>
          ))
        ) : (
          <p>No Results Found</p>
        )}
      </datalist>
    </Fragment>
  );
};

const mapStateToProps = (state) => {
  return {
    filterResults: selectPosts(state.posts.items, state.filters),
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text))
});

export default connect(mapStateToProps, mapDispatchToProps)(Autocomplete);