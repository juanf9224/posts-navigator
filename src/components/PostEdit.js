import React, { useState } from "react";
import { connect } from "react-redux";

import { editPostAction } from '../actions/posts';
import { setSelectedPost } from '../actions/selected-post';

const PostEdit = ({ setIsEdit, selectedPost, editPostAction, setSelectedPost }) => {
  const [postTitle, setPostTitle] = useState(selectedPost.title);
  const [postBody, setPostBody] = useState(selectedPost.body);

  const onSaveHandler = () => {
    const update = { title: postTitle, body: postBody };
    editPostAction(selectedPost.id, update );
    setSelectedPost({ ...selectedPost, ...update });
    setIsEdit(false);
  }

  return (
    <div className="card-layout">
      <div className="card-layout__content">
        <div className="card-layout__content__title">
          <input
            type="text"
            value={postTitle}
            onChange={(e) => setPostTitle(e.target.value)}
          ></input>
        </div>
        <div className="card-layout__content__body">
          <textarea
            value={postBody}
            onChange={(e) => setPostBody(e.target.value)}
          ></textarea>
        </div>
        <div className="card-layout__content__footer">
          <button className="btn btn-primary" onClick={onSaveHandler}>
            Save
          </button>
        </div>
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch, props) => ({
  editPostAction: (id, update) => dispatch(editPostAction(id, update)),
  setSelectedPost: (update) => dispatch(setSelectedPost(update)),
});

const mapStateToProps = (state) => {
  return {
    selectedPost: state.selectedPost,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(PostEdit);
