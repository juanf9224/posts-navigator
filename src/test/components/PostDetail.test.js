import React from "react";
import { shallow } from "enzyme";

import { PostDetail } from "../../components/PostDetail";
import posts from '../fixtures/posts';

describe("PostDetail component test suite", () => {
  let wrapper, setIsEdit, selectedPost;

  beforeEach(() => {
    setIsEdit = jest.fn();
    selectedPost = posts[0];
    wrapper = shallow(
      <PostDetail
        selectedPost={selectedPost}
        setSelectedPost={setIsEdit}
      />
    );
  });

  test("it should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
