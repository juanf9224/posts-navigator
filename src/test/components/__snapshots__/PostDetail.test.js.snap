// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetail component test suite it should match snapshot 1`] = `ShallowWrapper {}`;
